### DISCUSSION TITLE
Adding kind of Set and Map

### DISCUSSION THREAD

**Discussion by fkhadra** [pmndrs/valtio#317] [Ideas]
Hey, I use an abstraction layer to mimic Set and Map with Valtio. 
[![Edit frosty-rgb-eohxf](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/frosty-rgb-eohxf?fontsize=14&hidenavigation=1&theme=dark)

I was wondering if you think it's worth it to add that to the library. In that case, I could prepare a PR :). Otherwise, I can publish a library and mention it in the wiki. WDYT?

Thank you

**Comment by dai-shi** [comment]
Wow, this looks interesting...
You might have seen my comment somewhere. I don't plan to add Map/Set support in core.
But, your approach can be done on top of the public core API. It's actually acceptable and valuable.
Yeah, please try a PR to add functions in `valtio/utils`. If it's not accepted for some reason, then you can publish it as a third-party library.

I have a big request for this feature. It would be super nice to pass a Map for initial data.

```js
import { proxyMap } from 'valtio/utils'

const key1 = {}
const key2 = {}

const state = proxy({
  map: proxyMap(new Map([[key1, 'a']])),
})
state.map.set(key2, 'b')
```

We should also support non-string keys anyway, because that's the purpose of Map.
Hm, but then we need to copy Map on mutation to properly work as snapshot, which is important in React. I think it's worth the cost.

There will be more to consider. We would need lots of tests. Anyway, I'm excited!
